### PR TITLE
test: improve edge-case coverage for merge flows

### LINKED ISSUE #182: Improve test coverage for edge cases in transaction merge and submission flows
## Problem

Current test coverage for transaction merge and submission flows does not fully cover several important edge cases, such as:

* nil transaction inputs
* conflicting namespace writes
* duplicate endorsements
* context cancellation scenarios

This can allow subtle bugs or regressions to go undetected, especially in failure scenarios.

## Proposed Solution

Add targeted test cases to cover these edge conditions, ensuring that:

* invalid inputs are handled correctly
* conflicting or duplicate data is validated properly
* submission behavior is robust under cancellation scenarios

## Scope

* Add new test cases only
* No changes to production logic

## Benefits

* Improved reliability and robustness
* Better protection against regressions
* Clear validation of edge-case behavior

### PR DESCRIPTION
**Pull request by sachin9058**
#### Type of change

* Test update
* Improvement (improves reliability through better edge-case coverage)

#### Description

This PR enhances test coverage for transaction merging logic by adding targeted edge-case scenarios.

For Merge functionality:

* Added tests for nil transaction handling
* Covered conflicting namespace writes
* Verified duplicate endorsement deduplication
* Extended validation for multi-namespace scenarios and sorting behavior

These additions improve robustness by ensuring important failure paths and edge cases are properly validated, while keeping the test suite concise and non-redundant.

#### Additional details (Optional)

All tests pass locally using `go test ./...`.
No existing functionality was modified; changes are limited to test coverage improvements.

#### Related issues

Closes #182

### CODE REVIEW

**Comment by mbrandenburger** [comment]
Thank you @sachin9058 for creating this PR. Could you please create an issue, explain what the problem you are trying to solve, and then link this PR to your issue. Thank you

**Comment by sachin9058** [comment]
> Thank you @sachin9058 for creating this PR. Could you please create an issue, explain what the problem you are trying to solve, and then link this PR to your issue. Thank you

@mbrandenburger Hey I have created issue related to this pr. U can take a look on #182

**Comment by sachin9058** [comment]
@mbrandenburger Hey can u take a look on this wheather it looks good to u or require some more refinements or improvements ???

File: tools/fxconfig/internal/transaction/merge.go [hunk 1]
```diff
@@ -52,6 +52,7 @@ func sortEndorsementsByMspID(merged *applicationpb.Tx) {
 	}
 }
 
+// TODO: Add test coverage for edge cases in validation.
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge.go]:
```suggestion
```

File: tools/fxconfig/internal/app/endorse_test.go [hunk 2]
```diff
@@ -32,11 +32,15 @@ func (m *testSigningIdentity) Sign(_ []byte) ([]byte, error) {
...
-func (*testSigningIdentity) GetPublicVersion() msp.Identity                 { return nil } //nolint:ireturn
+func (*testSigningIdentity) SerializeWithIDOfCert() ([]byte, error) { return []byte{}, nil }
+func (*testSigningIdentity) Serialize() ([]byte, error)             { return []byte{}, nil }
+func (*testSigningIdentity) GetCertificatePEM() ([]byte, error)     { return nil, nil }
+func (*testSigningIdentity) GetIdentifier() *msp.IdentityIdentifier { return nil }
+
+//nolint:ireturn // interface return required by msp.SigningIdentity
+func (*testSigningIdentity) GetPublicVersion() msp.Identity {
+	return nil
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/endorse_test.go]: Not sure if we need these changes here.

File: tools/fxconfig/internal/app/submit_test.go [hunk 3]
```diff
@@ -25,7 +25,18 @@ type mockOrdererClient struct {
...
+	_ msp.SigningIdentity,
+	txID string,
+	_ *applicationpb.Tx,
+) error {
+	if txID == "" {
+		return errors.New("txID is required")
+	}
+	if err := ctx.Err(); err != nil {
+		return err
+	}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: we should not change this mock, instead it can return any error via `broadcastErr`

File: tools/fxconfig/internal/app/submit_test.go [hunk 4]
```diff
@@ -156,6 +167,36 @@ func TestSubmitTransaction_ContextCancelled(t *testing.T) {
...
+
+	a := &AdminApp{
+		MspProvider:     makeMSPProvider(&testSigningIdentity{}, nil),
+		OrdererProvider: makeOrdererProvider(&mockOrdererClient{}, nil),
+	}
+
+	err := a.SubmitTransaction(t.Context(), "", someTx())
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "txID is required")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: This test does not test the behavior of AdminApp, as the ordererClient is responsible to check the txID. Thus, this test only tests mock behavior - let's remove it.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 5]
```diff
@@ -110,6 +110,45 @@ func TestMerge_ErrorCases(t *testing.T) {
...
+}
+
+func TestMerge_NilTransaction(t *testing.T) {
+	t.Parallel()
+
+	result, err := Merge([]*applicationpb.Tx{nil})
+	require.Error(t, err)
+	require.Nil(t, result)
+	require.Contains(t, err.Error(), "at least two transactions required")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: I believe this test does not test the behavior of having actual nil transactions, but instead we run into the check how many transactions are provided. 

The interesting test would be what happens if we run:
```go
result, err := Merge([]*applicationpb.Tx{nil,nil})
result, err := Merge([]*applicationpb.Tx{nil,nil,nil})
```

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @sachin9058 for working on this. Good catching identifying some missing test cases. A few more comments below.

**Comment by sachin9058** [comment]
@mbrandenburger  Thanks for the feedback!

I’ve made the suggested updates:

* Reverted mock changes and removed validation from mocks
* Removed the empty txID test since it did not reflect AdminApp behavior
* Updated nil transaction tests to cover cases like [nil, nil] and [nil, nil, nil]
* Cleaned up unrelated changes

Please let me know if anything else should be refined.

**Comment by sachin9058** [comment]
@mbrandenburger hey can u review it again ??? and i have followed your feedback and make some changes...

**Comment by mbrandenburger** [comment]
Thank you @sachin9058 can you please go through my comments and see that you have remove unneccary changes as suggested. It seems there are still unneeded changes here.

**Comment by sachin9058** [comment]
@mbrandenburger Thanks for the feedback!

I’ve rechecked the changes and ensured that only relevant test updates remain in the PR. The signing identity and other test helpers are now identical to upstream, and the diff is limited to submit_test.go and merge_test.go.

Please let me know if anything else should be refined.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 6]
```diff
@@ -110,6 +110,56 @@ func TestMerge_ErrorCases(t *testing.T) {
...
 }
 
+func TestMerge_EmptyInput(t *testing.T) {
+	t.Parallel()
+
+	result, err := Merge([]*applicationpb.Tx{})
+	require.Error(t, err)
+	require.Nil(t, result)
+	require.Contains(t, err.Error(), "at least two transactions required")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: It seems that this case is already covered by `TestMerge_ErrorCases`

File: tools/fxconfig/internal/app/submit_test.go [hunk 7]
```diff
@@ -156,6 +156,23 @@ func TestSubmitTransaction_ContextCancelled(t *testing.T) {
...
+	}
+
+	ctx, cancel := context.WithTimeout(t.Context(), time.Nanosecond)
+	defer cancel()
+	<-ctx.Done()
+
+	err := a.SubmitTransaction(ctx, "tx-1", someTx())
+	require.Error(t, err)
+	require.ErrorIs(t, err, context.DeadlineExceeded)
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/app/submit_test.go]: I have the feeling that this test does not test the timeout itself; regardless of the ctx, the orderer mock always returns `context.DeadlineExceeded`. Therefore, it does not add any additional coverage to the already existing path in `TestSubmitTransaction_ContextCancelled`.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 8]
```diff
@@ -110,6 +110,56 @@ func TestMerge_ErrorCases(t *testing.T) {
...
+		require.Nil(t, result)
+	})
+
+	t.Run("three nil transactions", func(t *testing.T) {
+		t.Parallel()
+
+		result, err := Merge([]*applicationpb.Tx{nil, nil, nil})
+		require.Error(t, err)
+		require.Nil(t, result)
+	})
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: even though I suggested this one, the test with three nil transactions does not add any additional coverage here. We can drop it .. my bad :D

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @sachin9058 for last cleanup commits on this PR. Good work - here a few more comments. Can you also please update the PR description (remove "Added validation for empty transaction ID" as it's not update to date anymore). Thanks

**Comment by sachin9058** [comment]
@mbrandenburger Thanks for the feedback!

I’ve made the requested updates:

* Removed the redundant empty input test
* Removed the timeout test as it did not add additional coverage
* Removed the three nil transactions test
* Updated the PR description to reflect the current changes
* Aligned the issue description with the final scope of the PR

Please let me know if anything else should be refined.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 9]
```diff
@@ -110,6 +110,39 @@ func TestMerge_ErrorCases(t *testing.T) {
...
+func TestMerge_NilTransaction(t *testing.T) {
+	t.Parallel()
+
+	t.Run("two nil transactions", func(t *testing.T) {
+		t.Parallel()
+
+		result, err := Merge([]*applicationpb.Tx{nil, nil})
+		require.Error(t, err)
+		require.Nil(t, result)
+	})
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: No need to have this as a sub test here.

File: tools/fxconfig/internal/transaction/merge_test.go [hunk 10]
```diff
@@ -110,6 +110,39 @@ func TestMerge_ErrorCases(t *testing.T) {
...
+	tx2.Namespaces[0].ReadWrites = []*applicationpb.ReadWrite{{
+		Key:   []byte("asset-1"),
+		Value: []byte("value-b"),
+	}}
+
+	result, err := Merge([]*applicationpb.Tx{tx1, tx2})
+	require.Error(t, err)
+	require.Nil(t, result)
+	require.Contains(t, err.Error(), "content mismatch")
+}
```
Comment (mbrandenburger) [inline tools/fxconfig/internal/transaction/merge_test.go]: should this be better a subtest of `TestMerge_ErrorCases`? WDYT?

**Review by mbrandenburger — CHANGES_REQUESTED** [review]
Thank you @sachin9058 for working on this. It seems the PR description is still not quite right. It says:

> For SubmitTransaction:
> 
> - Added tests for context cancellation handling
> - Covered error scenarios for MSP and orderer providers

These changes are not part of the PR (anymore). 
More comments below.

**Comment by sachin9058** [comment]
@mbrandenburger Thanks for the feedback!

I’ve made the final updates:

* Fixed the lint issue by aligning with the repository’s lint rules
* Cleaned up the test structure as suggested
* Ensured the PR now only contains relevant and non-redundant test cases

All checks are passing now. Please let me know if anything else should be refined.

**Review by mbrandenburger — APPROVED** [review]
Thank you @sachin9058

**Comment by sachin9058** [comment]
@mbrandenburger Thanks a lot for the review and guidance throughout this PR I really appreciate it!

I learned a lot about keeping changes minimal and aligned with the project’s design. Looking forward to contributing more.
